### ISSUE TITLE
improve OverloadMethodsDeclarationOrder to optionally group static methods separately from instance methods

### ISSUE THREAD

**Issue by ilanKeshet** [checkstyle/checkstyle#3698]
Consider the following code:
```java
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
      <module name="OverloadMethodsDeclarationOrder"/>
    </module>
</module>

$ cat Test.java 
public class Test {
    void foo() { /* code */ }
    void foo(String baz) { /* code */ }
    void bar() { /* code */ }
    static void foo() { /* code */ } // detected as out of order - point of issue
}

$ java -jar checkstyle-7.4-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/rivanov/nfs/tests/Test.java:5: Overload methods should not be split. Previous overloaded method located at line '3'. [OverloadMethodsDeclarationOrder]
Audit done.
Checkstyle ends with 1 errors.
```
---------------
Currently `OverloadMethodsDeclarationOrder` only considers the method name and does not take into account other information.

The **suggested improvement** is to allow a user to choose whether `static` methods should or shouldn't be considered as overloading instance methods.

```xml
    <module name="OverloadMethodsDeclarationOrder">
        <property name="separateStatic" value="true" />
    </module>
```
The default should obviously `separateStatic=false` to maintaining the current functionality `<module name="OverloadMethodsDeclarationOrder"/>`

Should this feature be expected, i'd be happy to contribute it's implementation.

--------------
Note: 
- This issue, is similar, but not a duplicate of [OverloadMethodsDeclarationOrder reports an error when static and instance method not together: opportunity for improvement? #2652](https://github.com/checkstyle/checkstyle/issues/2652) as it treats current implementation as a buggy, while this issue proposes to provide the user with the flexibility of configuration
- [Apply OverloadMethodsDeclarationOrder check for the same access modificators only #3235](https://github.com/checkstyle/checkstyle/issues/3235), *requesting a similar sort of functionality scoped by method visibility*, and this issue can be implemented together

**Comment by romani** [comment]
#2652 is absolutely the same, but ssue was closed as author did not provide me more details and more examples of usage. I also raise for him my concern with visibility (as it was later raised in #3235) that was not answered.

Right now we have only 3 issues that looks like design solutions become more general.
You want to split by "static" , #3235 want to split by visibility.
All of them are modifiers - http://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-MethodModifier
So we could make option like "groupBySameModifiers" to force to group methods with the same modifiers.

Example for `groupBySameModifiers=true`:
```
public class Test {
    void foo() { /* code */ } //OK
    void foo(String baz) { /* code */ } //OK
    void bar() { /* code */ }
    static void foo() { /* code */ } // OK
    static void foo(String str) { /* code */ } // OK
}

public class Test {
    static private void foo() { /* code */ } // OK, new set of modifiers - "static private" man separate group
    void foo() { /* code */ } //OK
    void foo(String baz) { /* code */ } //OK
    void bar() { /* code */ }
    static void foo() { /* code */ } // OK
}

public class Test {
    private void foo() { /* code */ } //OK
    void bar() { /* code */ }
    public void foo(String baz) { /* code */ } //OK
    public void foo(String baz) { /* code */ } //OK
    private void foo(int i) { /* code */ } // VIOLATION
    private final void foo(int i) { /* code */ } // OK, "private final " is separate group
}

public class Test {
    protected void foo() { /* code */ } //OK
    void bar() { /* code */ }
    protected abstract void foo() { /* code */ } //OK - separate group by modifiers
    private void foo(int i) { /* code */ } // OK
    private final void foo(int i) { /* code */ } // OK, "private final " is separate group
}
```

@ychulovskyy , @supercobra, @LoPoBo , @rnveach, @MEZk  please share your ideas and expectations to satisfy all needs by the same solution (if possible)

**Comment by romani** [comment]
@ilanKeshet , can at least you share thoughts ?

**Comment by ilanKeshet** [comment]
Yes, thank you for the reminder and Sorry for the late reply. 

I'm not sure it's an all or nothing kind of choice.
i.e. `groupBySameModifiers=true` is too broad, forcing the user to choose between grouping by method name only (as today) Or grouping by too many dimensions, and rendering the Rule kind of useless.  

I would like, to have at least two switches 
* `groupByVisibilityModifiers=true` 
* `groupByStaticModifier=true` 

Not sure if needed, but in theory users might want the ability to separately control: 
* `groupBySynchronicityModifier=true` 
* `groupByFinalityModifier=true` 
* `groupByImplementationModifier=true` // group by `abstract` , `default`  and `native`

**Comment by rnveach** [comment]
@romani I agree with @ilanKeshet on just wanting this for `static` alone. I too prefer similar code in my project.

````
public class Test {
    static private void foo() {  } // OK, new set of modifiers - "static private" man separate group
    void foo() {  } //OK
    void foo(String baz) {  } //OK
    void bar() {  }
    static void foo() {  } // VIOLATION, static foo should be next to static private foo.
}
````

I question the use of `true` or `false` switches as he demonstrated as it is not clear which grouping takes priority, `static` or `visibility` if both are on, but that just depends on if someone wants that much customization.

**Comment by ilanKeshet** [comment]
@rnveach , good point.  I thought of the switches as being additive. where enabling a switch allows for a group with similar name to be split. 

also, perhaps a the switch name would be better understood if writes as `allowSplitByVisibility=true` 
and `allowSplitByStatic=true` 

**some example of additive switches use:**

assuming all flags are `false` : 
````
public class Test {
    public void foo() {  } //OK
    public void bar(String baz) {  } //OK
    private void foo(String baz) {  } //VIOLATION private foo(String) should be next to foo()
    void bar() {  } // VIOLATION bar should be next to bar(String)
    static void foo() {  } // VIOLATION, static foo should be next to foo(String).
}
````

assuming all flags are `false` except `groupByStaticModifier=true` : 
````
public class Test {
    public void foo() {  } //OK
    private void bar(String baz) {  } //OK
    private void foo(String baz) {  } //VIOLATION private foo(String) should be next to foo()
    void bar() {  } // VIOLATION bar should be next to bar(String)
    static void foo() {  } // OK, static foo is it's own group. 
}
````

assuming all flags are `false` except `groupByStaticModifier=true` and `groupByVisibilityModifiers=true` : 
````
public class Test {
    public void foo() {  } //OK
    private void bar(String baz) {  } //OK
    private void foo(String baz) {  } //OK private foo(String) is it's own group
    void bar() {  } // OK package scope bar is it's own group
    static void foo() {  } // OK, static foo is it's own group. 
}
````

**Comment by MEZk** [comment]
@rnveach , @romani , @ilanKeshet 
The main point of the issue was the idea of adding a new option called 'separateStatic'.  However, the option is too strict for me. I like @romani's proposal (groupBySameModifiers) as I prefer to group methods by modifiers. This is what I expect if groupBySameModifiers=true :

```java
public class A {
  public void foo1() {} // OK
  public void foo2() {} // OK
  public void foo1(String param1) {} // VIOLATION
  public void foo3() {} // OK
  public static void foo3() {} // VIOLATION
  public void foo5() {} // OK
  private void foo5(int a) {} // VIOLATION
  public void foo5(String a) {} // OK
}
```
```java
public class B {
  public void foo1() {} // OK
  public final void foo2(String a) {] // VIOLATION
  public void foo1(String a, String b) {} // OK
}
```

```java
public class C {
  public void foo1() {} // OK
  public void foo1(String param1) {} // OK
  public void foo2() {} // OK
  public void foo2(int a) {} // OK

  public static void foo4() {} // OK
  public static void foo5() {} // OK
  public static void foo5(String param1) {} // OK
  public static void foo5(String param1, String param2) {} // OK

  private void foo6() {} // OK
  private void foo6(String param1) {} // OK
  private void foo7() {} // OK
}
```
-------------------------------------------------
groupBySameModifiers=true 
```java
public class Test {
    void foo() { /* code */ } // OK
    void foo(String baz) { /* code */ } // OK
    void bar() { /* code */ } // OK
    static void foo() { /* code */ } // OK
}
```

groupBySameModifiers=false 
```java
public class Test {
    void foo() { /* code */ } // OK
    void foo(String baz) { /* code */ } // OK
    void bar() { /* code */ } // OK
    static void foo() { /* code */ } // VIOLATION
}
```
------------------------------
Additional options (allowSplitByStatic, allowSplitByVisibility, groupByStaticModifier, groupByVisibilityModifiers) will make the logic too complicated.

**Comment by MEZk** [comment]
After discussion with @romani we came to a conclusion that the examples and option (groupBySameModifiers) was difficult to understand. It will be mislead users of the check.

**Comment by romani** [comment]
I removed comment of MEZk, as it produce only even more mess, and show that my initial model was far from good.

I do not like approach of @ilanKeshet , as it creates too much properties, support of such properties is known problem. So lets consider another approaches.

Here is another idea from me:
property "ModifiersThatGreateGroup" = list of String.
Examples : 
`ModifiersThatGreateGroup` = "static" - all static methods should be separate (should cover request of this issue), 
"ModifiersThatGreateGroup" = "public, private, protected" - if any of this modifier is met it should be in separate group for overloads (should cover request to split by visibility).
for smb who want to go crazy 
"ModifiersThatGreateGroup" = "static final , public native, private static, protected abstract" - it will keep method with "static final" , "public native", "private static", "protected abstract" is separate groups, only if both modifiers are on method , order of modifiers should not matter.

I will do examples of code and configs a bit later........

**Comment by romani** [comment]
@ychulovskyy , @supercobra, @LoPoBo , @rnveach, @MEZk , @ilanKeshet please review and be welcome with feedback.

**Comment by ilanKeshet** [comment]
@romani This is a good approach, I personally like it very much as it allows fine grain control over which groups to form.
However I think it is far from simple, and would result the same type of usage pattern of *Copy & Paste* configuration. Same as [DescendantToken](http://checkstyle.sourceforge.net/config_misc.html#DescendantToken) where the complexity of usage is such that (at least in my limited view) users only use the documented examples, making them effectively property flags. 

Initially I played around with such an idea before reporting the issue, but decided to quash it as I got hung up on using, what seems to be the convention of using [TokenTypes](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java) constants and I couldn't find one for `package` scope. 

**Comments and Questions:** 
1. The regex you suggest seem to relay on the the modifiers order. I must say i don't immediately understand what should be the difference in group formation between `^public .*` and `public` other than that in the first example the public must be the first modifier. perhaps another way to phrase it would be to ask *what does `.*` mean?* 
2. If I understand correctly the backward compatible path is achieved by always having *default group* , is that right? what if users what want to disable it? 
3. what about interfaces.
3.1. Obviously in java 8 the `default`  keyword also needs to be supported. 
3.2. Would grouping behave differently if `public abstract` is written or omitted? 
4. `package` how would `^package` behave given that the literal word isn't really there? 
5. could a method belong to different groups, for example `^public .*` and `.* static` ?

**Comment by ychulovskyy** [comment]
@romani I would concentrate on KISS.
//ModifiersThatGreateGroup = "protected, private, package" is easy to write and understand.
And it resolves my problem.

Let's think about another simple case: keep static methods together, keep public method together: 
````
//ModifiersThatGreateGroup = "static, public"
public class Test {
    public void foo(String f) {  } //OK, public group
    public void bar(String f) {  } 
    public static void foo() {  } //Is it public or static group? both? neither?
}
````
Does  ````public static void foo()```` belong to public group or static group? Both groups or neither?
Should I write "public" or "public .*" if I want to group by public modifier?
I don't know good solution, so suggest to document strategy of assigning methods to groups.

**Comment by romani** [comment]
@ilanKeshet , thanks a lot for detailed review.
Here are the answers.

>However I think it is far from simple, and would result the same type of usage pattern of Copy & Paste configuration.

advanced tuning, demands advanced understanding of configuration. User will have examples of most common requests to use. Most users should not have extra complicated ordering, just split for visibility and static is more than enough.

> Same as DescendantToken where the complexity of usage is such that

That is why we have WARNING on usage of it. It is for advanced and very demanding users . Just a tool, all responsibility is on user.

> what seems to be the convention of using TokenTypes constants and I couldn't find one for package scope.

package scope is virtual scope(default) term in jdk specification only, no java language literal for it - so no TokenTypes for it.
Yes, usage of TokenTypes is preferable, but real life is far from that simple, we already use a lot of custom types. But try to use RegExp as much as possible before new types introduction.

> 1. The regex you suggest seem to relay on the the modifiers order.

yes, is user use [ModifierOrder](http://checkstyle.sourceforge.net/config_modifier.html#ModifierOrder) it would ease his configuration, but we try to avoid any dependencies between Checks. All Checks should be independent to do validation correctly alone.  

> I must say i don't immediately understand what should be the difference in group formation between ^public .* and public ...  perhaps another way to phrase it would be to ask what does .* mean?

Yes, but that would be documented, that we provide user RegExp power, all responsibility of configuration is on user.

> If I understand correctly the backward compatible path is achieved by always having default group , is that right?

yes, I think by default this property should be empty list, so no special groups (all matches to default).

>  what if users what want to disable it?

no answer for now. I think in this case user need to clearly define all regexps to match what he needs and default will be just group that is in mind of user but not in configuration. Not ideal solution but I think it is fine for now. If the will be storm of request to fix this, we will have enough user cases to make extension.

> Obviously in java 8 the default keyword also needs to be supported

We give user RegExp, default is real text in code so, user will decide himself on grouping by it or not.
We should add only "package" as it doe snot exists in code text but required for full visibility grouping.
 
> Would grouping behave differently if public abstract is written or omitted?

Yes, it will.
We will just put all modifiers in one String and match by RegExps, all responsibility in on user.
Right now Check does not support differentiation of TokenTypes to run at. So it might be good another issue to allow user to have different configs for classes and interfaces.

> package how would ^package behave given that the literal word isn't really there?

good catch, we might just make it in documentation clear that will append that modifier to the end.
We will describe an algorithm of creation of modifiers string and how it matches. As we give user tool , we have to be transparent to him.

> could a method belong to different groups, for example ^public .* and .* static ?

My position right now is to follow users order of regexps in config, it will give user more power and in the same time match process is strict.

**Comment by romani** [comment]
@ychulovskyy , thanks for feedback.

> ModifiersThatGreateGroup = "protected, private, package" is easy to write and understand.

it is not ideal, it force you to understand that all `public` will stay in default group. But yes, config looks simple.

> keep static methods together, keep public method together

good use case.
Config should be: `modifiersThatGreateGroup` = "static, public" (exactly as you defines, with nuance that "static" is first)
user defined string, will be split by `,` , then for each part we create RegExp, so we will have array of [RegExp:"static", RegExp:"public"]. Match to RegExps will follow user order of RegExps. We will do first match search.

> Does public static void foo() belong to public group or static group? Both groups or neither?

to first group that match. All responsibility is on user.

> Should I write "public" or "public .*" if I want to group by public modifier?

As it is RegExp, just "public" is enough.

> I don't know good solution, so suggest to document strategy of assigning methods to groups.

yes, this property should be documented in details with examples and clear description of how we consume user config and do matching. Looks like it will not be overcomplicated property. Unfortunately it will not be obvious without reading documentation.

**Comment by romani** [comment]
@ilanKeshet ,@ychulovskyy ,
Please confirm that you see no problems with design for now and we could convert it to issue with exact specification .

 @ilanKeshet , do you still have ability to help us implement this ?

**Comment by ychulovskyy** [comment]
@romani confirmed

**Comment by ilanKeshet** [comment]
@romani confirmed, looks good to go.

Yes, my offer to help implementing this issues still stands, I could start Friday (Feb 3)

**Comment by romani** [comment]
closed in favor of #3770
